src/compiler/checker.ts

```diff
@@ -3581,6 +3582,9 @@ module ts {
             }
         }
 
+        // The noSubtypeReduction flag is there because it isn't possible to always do subtype reduction. The flag
+        // is true when creating a union type from a type node and when instantiating a union type. In both of those
+        // cases subtype reduction has to be deferred to properly support recursive union types.

```

Comment:
I think it'd be useful to put an example of each of the two cases, either here or at the relevant calls to getUnionType